core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -374,12 +374,6 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
-    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
-    if (master == "yarn" && deployMode == "cluster" && !_conf.contains("spark.yarn.app.id")) {
```

Comment:
Hmm, this was introduced so that people don't try `new SparkContext(new SparkConf().setMaster("yarn-cluster"))` and then complain when they get really odd error messages.

Is there another way to do this check?